app/components/AudioPlayer/AudioPlayer.tsx

```diff
@@ -0,0 +1,163 @@
+import React, { useEffect, useRef, useState, useCallback } from 'react';
+import { Box, IconButton } from '@mui/material';
+import { styles } from './AudioPlayer.styles';
+import AudioPlayerPopover from './AudioPlayerPopover/AudioPlayerPopover';
+
+type AudioPlayerProps = {
+  src: string;
+  trackName: string;
+  loop?: boolean;
+  autoplay?: boolean;
+  className?: string;
+};
+
+export default function AudioPlayer({
+  src,
+  trackName,
+  loop = false,
+  autoplay = false,
+  className,
+}: AudioPlayerProps) {
+  const audioRef = useRef<HTMLAudioElement | null>(null);
+  const buttonRef = useRef<HTMLButtonElement | null>(null);
+
+  const [isPlaying, setIsPlaying] = useState(false);
+  const [currentTime, setCurrentTime] = useState(0);
+  const [duration, setDuration] = useState(0);
+  const [anchorEl, setAnchorEl] = useState<HTMLButtonElement | null>(null);
+  const [error, setError] = useState<string | null>(null);
+
+  const stickHeights = [7, 19, 30, 13, 22, 7];
+
+  useEffect(() => {
+    const audio = audioRef.current;
+    if (!audio) return;
+
+    const updateProgress = () => {
+      setCurrentTime(audio.currentTime);
+    };
+
+    const updatePlayState = () => {
+      setIsPlaying(!audio.paused);
+    };
+
+    const loadDuration = () => {
+      if (!isNaN(audio.duration) && audio.duration > 0) {
+        setDuration(audio.duration);
+      }
+    };
+
+    const handleEnded = () => {
+      setIsPlaying(false);
+      setAnchorEl(null);
+    };
+
+    const handleError = () => {
+      setError('Error loading audio file.');
+    };
+
+    const handleLoadedMetadata = () => {
+      loadDuration();
+      setError(null);
+    };
+
+    audio.addEventListener('timeupdate', updateProgress);
+    audio.addEventListener('play', updatePlayState);
+    audio.addEventListener('pause', updatePlayState);
+    audio.addEventListener('ended', handleEnded);
+    audio.addEventListener('loadedmetadata', handleLoadedMetadata);
+    audio.addEventListener('error', handleError);
+
+    if (audio.readyState >= 1) {
+      loadDuration();
+      setError(null);
+    }
+
+    if (autoplay) {
+      audio.play().catch(() => {});
+    }
+
+    return () => {
+      audio.removeEventListener('timeupdate', updateProgress);
+      audio.removeEventListener('play', updatePlayState);
+      audio.removeEventListener('pause', updatePlayState);
+      audio.removeEventListener('ended', handleEnded);
+      audio.removeEventListener('loadedmetadata', handleLoadedMetadata);
+      audio.removeEventListener('error', handleError);
+    };
+  }, [autoplay, src]);
+
+  const togglePlay = useCallback((): void => {
```

Comment:
Why don't you have common principles for callback typing? Somewhere you have types for callbacks, but sometimes not. I suggest you not type this callback, it's redundant.
```suggestion
  const togglePlay = useCallback(() => {
```

Reply:
Ok, I`ll fix it